File: source/frontend/src/lib/recurringTransaction.ts
import { useQuery } from '@tanstack/react-query'

import { api } from './api'
import { accountQueryKeys } from './accountHistory'
import { authQueryKeys } from './auth'
import { useInvalidatingMutation } from './mutation'
import type { CategoryKey, TransactionType } from './transaction'

export const RECURRENCE_FREQUENCIES = ['MONTHLY', 'WEEKLY'] as const

export type RecurrenceFrequency = (typeof RECURRENCE_FREQUENCIES)[number]

export interface RecurringTransactionRead {
  id: number
  account_id: number
  amount: number
  purpose: string | null
  other_party: string | null
  transaction_type: TransactionType | null
  category: CategoryKey | null
  note: string | null
  frequency: RecurrenceFrequency
  day_of_month: number | null
  day_of_week: number | null // 0 = Monday … 6 = Sunday
  next_run_date: string // ISO yyyy-mm-dd
}

export interface RecurringTransactionCreatePayload {
  amount: number
  purpose?: string | null
  other_party?: string | null
  transaction_type?: TransactionType | null
  category?: CategoryKey | null
  note?: string | null
  frequency: RecurrenceFrequency
  day_of_month?: number | null
  day_of_week?: number | null
  book_immediately: boolean
}

// Same shape as the create payload, minus the one-off "book today" flag.
export type RecurringTransactionUpdatePayload = Omit<
  RecurringTransactionCreatePayload,
  'book_immediately'
>

export const recurringQueryKeys = {
  list: (accountId: number) => ['account', accountId, 'recurring'] as const,
}

export function useRecurringTransactions(accountId: number) {
  return useQuery({
    queryKey: recurringQueryKeys.list(accountId),
    queryFn: () => api<RecurringTransactionRead[]>(`/account/${accountId}/recurring-transactions`),
  })
}

export function useCreateRecurringTransaction(accountId: number) {
  return useInvalidatingMutation({
    mutationFn: (payload: RecurringTransactionCreatePayload) =>
      api<RecurringTransactionRead>(`/account/${accountId}/recurring-transactions`, {
        method: 'POST',
        body: payload,
      }),
    invalidate: [
      recurringQueryKeys.list(accountId),
      accountQueryKeys.history(accountId),
      authQueryKeys.me,
    ],
  })
}

export function useUpdateRecurringTransaction(accountId: number) {
  return useInvalidatingMutation({
    mutationFn: ({
      recurringTransactionId,
      payload,
    }: {
      recurringTransactionId: number
      payload: RecurringTransactionUpdatePayload
    }) =>
      api<RecurringTransactionRead>(
        `/account/${accountId}/recurring-transactions/${recurringTransactionId}`,
        { method: 'PATCH', body: payload },
      ),
    invalidate: [recurringQueryKeys.list(accountId)],
  })
}

export function useDeleteRecurringTransaction(accountId: number) {
  return useInvalidatingMutation({
    mutationFn: (recurringTransactionId: number) =>
      api<void>(`/account/${accountId}/recurring-transactions/${recurringTransactionId}`, {
        method: 'DELETE',
      }),
    invalidate: [recurringQueryKeys.list(accountId)],
  })
}

File: source/backend/api/schemas/transactions/recurring_transaction.py
import datetime

from pydantic import BaseModel, ConfigDict, Field, model_validator

from source.backend.models.transactions.recurrence_frequency import RecurrenceFrequency
from source.backend.models.transactions.transaction_category import TransactionCategory
from source.backend.models.transactions.transaction_type import TransactionType


class _RecurringTransactionFields(BaseModel):
    amount: float
    purpose: str | None = None
    other_party: str | None = None
    transaction_type: TransactionType | None = None
    category: TransactionCategory | None = None
    note: str | None = None

    frequency: RecurrenceFrequency
    day_of_month: int | None = Field(default=None, ge=1, le=31)
    day_of_week: int | None = Field(default=None, ge=0, le=6)

    @model_validator(mode="after")
    def _check_schedule(self) -> "_RecurringTransactionFields":
        if self.frequency == RecurrenceFrequency.MONTHLY:
            if self.day_of_month is None:
                raise ValueError(f"day_of_month is required for a {RecurrenceFrequency.MONTHLY.value} recurrence")
            if self.day_of_week is not None:
                raise ValueError(f"day_of_week must be omitted for a {RecurrenceFrequency.MONTHLY.value} recurrence")
        else:
            if self.day_of_week is None:
                raise ValueError(f"day_of_week is required for a {RecurrenceFrequency.WEEKLY.value} recurrence")
            if self.day_of_month is not None:
                raise ValueError(f"day_of_month must be omitted for a {RecurrenceFrequency.WEEKLY.value} recurrence")
        return self


class RecurringTransactionCreate(_RecurringTransactionFields):
    book_immediately: bool = False


class RecurringTransactionUpdate(_RecurringTransactionFields):
    pass


class RecurringTransactionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    account_id: int
    amount: float
    purpose: str | None
    other_party: str | None
    transaction_type: TransactionType | None
    category: TransactionCategory | None
    note: str | None
    frequency: RecurrenceFrequency
    day_of_month: int | None
    day_of_week: int | None
    next_run_date: datetime.date
